Fix race condition in reconnect on reticulum deploy

johnshaughnessy opened the pull request:
If `getReticulumMeta` does not resolve for more than 5 seconds, or if reticulum issues multiple `ret-deploy` events for some reason, the previous `retDeployReconnectInterval` would call `tryReconnect` multiple times, resulting in a race condition in the subsequent block. 

This change avoids the race condition. On the off-chance we receive multiple `ret-deploy` events in rapid succession, the handler for the first event can be interrupted so that the most recent event can be processed as soon as possible.

johnshaughnessy commented:
The following test shows the new code handling `ret-deploy` events one at a time:
```js
22:51:26.158 {
  // Simulate ret-deploy notifications from reticulum
  // Note: I shortened the delays between reconnecting to accelerate testing.

  // Correct combo
  window.onRetDeploy({event: "ret-deploy", ret_version: "1.0.20210323184939", ret_pool: "earth"})
  window.onRetDeploy({event: "ret-deploy", ret_version: "1.0.20210323184939", ret_pool: "earth"})

  // Some bad combos
  window.onRetDeploy({event: "ret-deploy", ret_version: "fake version", ret_pool: "earth"})
  window.onRetDeploy({event: "ret-deploy", ret_version: "1.0.20210323184939", ret_pool: "fake pool"})
  window.onRetDeploy({event: "ret-deploy", ret_version: "fake version", ret_pool: "fake pool"})
  window.onRetDeploy({event: "ret-deploy", ret_version: "1.0.20210323184939", ret_pool: "arbre"})

  // Correct combo
  window.onRetDeploy({event: "ret-deploy", ret_version: "1.0.20210323184939", ret_pool: "earth"})
}
```
```
22:51:26.199 [reconnect] Reticulum deploy detected v1.0.20210323184939 on earth. hub.js:1123:12
22:51:26.200 [reconnect] Attempting reconnect in 2 seconds. hub.js:1132:16
22:51:26.201 undefined
22:51:28.577 [reconnect] Reconnect in progress. Updated reticulum meta. hub.js:1152:12
22:51:28.577 Phoenix Socket URL: wss://pensive-dwarf.reticulum.io phoenix-utils.js:125:10
22:51:39.509 [reconnect] Reconnection successful. hub.js:1163:12
22:51:39.509 [reconnect] Reticulum deploy detected v1.0.20210323184939 on earth. hub.js:1123:12
22:51:39.510 [reconnect] Attempting reconnect in 2 seconds. hub.js:1132:16
22:51:40.791 [reconnect] Reconnect in progress. Updated reticulum meta. hub.js:1152:12
22:51:40.792 Phoenix Socket URL: wss://gallant-dwarf.reticulum.io phoenix-utils.js:125:10
22:51:51.335 [reconnect] Reconnection successful. hub.js:1163:12
22:51:51.336 [reconnect] Reticulum deploy detected vfake version on earth. hub.js:1123:12
22:51:51.336 [reconnect] Attempting reconnect in 2 seconds. hub.js:1132:16
22:51:52.912 [reconnect] Attempting reconnect in 2 seconds. (Attempt 2 of 4) hub.js:1132:16
22:51:54.593 [reconnect] Attempting reconnect in 4 seconds. (Attempt 3 of 4) hub.js:1132:16
22:51:58.122 [reconnect] Attempting reconnect in 4 seconds. (Attempt 4 of 4) hub.js:1132:16
22:52:02.243 [reconnect] Failed to reconnect to reticulum. hub.js:1148:14
22:52:02.244 [reconnect] Reticulum deploy detected v1.0.20210323184939 on fake pool. hub.js:1123:12
22:52:02.244 [reconnect] Attempting reconnect in 1 seconds. hub.js:1132:16
22:52:03.199 [reconnect] Attempting reconnect in 3 seconds. (Attempt 2 of 4) hub.js:1132:16
22:52:06.193 [reconnect] Attempting reconnect in 5 seconds. (Attempt 3 of 4) hub.js:1132:16
22:52:11.295 [reconnect] Attempting reconnect in 4 seconds. (Attempt 4 of 4) hub.js:1132:16
22:52:14.526 [reconnect] Failed to reconnect to reticulum. hub.js:1148:14
22:52:14.527 [reconnect] Reticulum deploy detected vfake version on fake pool. hub.js:1123:12
22:52:14.527 [reconnect] Attempting reconnect in 3 seconds. hub.js:1132:16
22:52:17.028 [reconnect] Attempting reconnect in 3 seconds. (Attempt 2 of 4) hub.js:1132:16
22:52:19.813 [reconnect] Attempting reconnect in 3 seconds. (Attempt 3 of 4) hub.js:1132:16
22:52:22.605 [reconnect] Attempting reconnect in 6 seconds. (Attempt 4 of 4) hub.js:1132:16
22:52:27.941 [reconnect] Failed to reconnect to reticulum. hub.js:1148:14
22:52:27.941 [reconnect] Reticulum deploy detected v1.0.20210323184939 on arbre. hub.js:1123:12
22:52:27.942 [reconnect] Attempting reconnect in 2 seconds. hub.js:1132:16
22:52:29.691 [reconnect] Attempting reconnect in 4 seconds. (Attempt 2 of 4) hub.js:1132:16
22:52:33.008 [reconnect] Attempting reconnect in 3 seconds. (Attempt 3 of 4) hub.js:1132:16
22:52:36.090 [reconnect] Attempting reconnect in 5 seconds. (Attempt 4 of 4) hub.js:1132:16
22:52:40.578 [reconnect] Failed to reconnect to reticulum. hub.js:1148:14
22:52:40.579 [reconnect] Reticulum deploy detected v1.0.20210323184939 on earth. hub.js:1123:12
22:52:40.579 [reconnect] Attempting reconnect in 3 seconds. hub.js:1132:16
22:52:42.963 [reconnect] Reconnect in progress. Updated reticulum meta. hub.js:1152:12
22:52:42.964 Phoenix Socket URL: wss://pensive-dwarf.reticulum.io phoenix-utils.js:125:10
22:52:53.559 [reconnect] Reconnection successful. hub.js:1163:12
```

johnshaughnessy commented on src/hub.js:
```diff
@@ -1115,46 +1113,86 @@ document.addEventListener("DOMContentLoaded", async () => {
+    retPhxChannel = await migrateChannelToSocket(retPhxChannel, socket);
+    await hubChannel.migrateToSocket(socket, createHubChannelParams());
+    authChannel.setSocket(socket);
+    linkChannel.setSocket(socket);
```
This section is left unchanged. If there are any retry loops within that need amending, they should be handled in a different PR.

netpro2k commented on src/hub.js:
```diff
@@ -1115,46 +1113,86 @@ document.addEventListener("DOMContentLoaded", async () => {
-    // On Reticulum deploys, reconnect after a random delay until pool + version match deployed version/pool
-    console.log(`Reticulum deploy detected v${deployNotification.ret_version} on ${deployNotification.ret_pool}`);
-    clearInterval(retDeployReconnectInterval);
+  const reconnect = async ({ ret_pool, ret_version }) => {
```
Unless I am missing something, this is not doing any sort of reconnecting, so the name is confusing. I think what it's doing is trying to find out if the version we just got a notification for is live, the re-connection happens later.

netpro2k commented on src/hub.js:
```diff
@@ -1115,46 +1113,86 @@ document.addEventListener("DOMContentLoaded", async () => {
+          }`
+        );
+        await sleep(delayMS);
+        // Reconnect until reticulum meta matches expected version and pool.
```
This comment made me think there would be a retry loop inside this function call. I honestly would probably just inline it, or move the loop into it, since its the core of what this loop is doing. Waiting for the version to be live.

netpro2k commented on src/hub.js:
```diff
@@ -1115,46 +1113,86 @@ document.addEventListener("DOMContentLoaded", async () => {
+  const migrateToNewReticulumServer = async ({ ret_version, ret_pool }) => {
+    console.log(`[reconnect] Reticulum deploy detected v${ret_version} on ${ret_pool}.`);
+
+    const maxAttempts = 4;
```
We may want to do more attempts than this. Not sure what the typical time between notification going out and servers being live, but I imagine it could be a bit, and we would really like to not time people out if we can avoid it.... maybe we need some sort of backoff?

johnshaughnessy replied:
Added more attempts now that there's an exit hatch. There was already an incremental backoff of `attempt * 3000` milliseconds. I adjusted this to `attempt * 5000`.

johnshaughnessy replied:
Added more attempts now that there's an exit hatch. There was already an incremental backoff of `attempt * 3000` milliseconds. I adjusted this to `attempt * 5000`.

netpro2k commented on src/utils/async-utils.js:
```diff
@@ -11,3 +11,9 @@ export const waitForDOMContentLoaded = function() {
   }
 };
+
+export const sleep = function(milliseconds) {
```
Kind of crazy we have an async-utils.js and don't have this heh

johnshaughnessy replied:
Not sure if it's any better than `await new Promise(res => setTimeout(res, 5000));`  :shrug:

johnshaughnessy replied:
Not sure if it's better than writing `await new Promise(res => setTimeout(res, 5000))` but :shrug:

netpro2k commented on src/hub.js:
```diff
@@ -1115,46 +1113,86 @@ document.addEventListener("DOMContentLoaded", async () => {
-    }, Math.floor(Math.random() * retReconnectMaxDelayMs));
-  };
+    return function onRetDeploy(deployNotification) {
+      notificationQueue.push(deployNotification);
```
Hmm, queuing these feels a bit odd. If we happen to do multiple deploys back to back the previous ones are now invalid and will never resolve, so we are guaranteed the reconnect attempt will fail, so it might make sense to short circuit them and save us some load.

netpro2k reviewed (commented):
Logic wise I think this seems good. Made some suggestions structure/naming wise and think we should talk through retry caps/delays, but otherwise lgtm

johnshaughnessy commented:
I updated the code and tested again. Notice that handling the first event is interrupted and the second is not processed at all.
```js
15:52:04.492 {
  window.onRetDeploy({event: "ret-deploy", ret_version: "fake version", ret_pool: "fake pool"})
  window.onRetDeploy({event: "ret-deploy", ret_version: "1.0.20210323184939", ret_pool: "fake pool"})
  window.onRetDeploy({event: "ret-deploy", ret_version: "1.0.20210323184939", ret_pool: "earth"})
}
15:52:04.545 [reconnect] Reticulum deploy detected vfake version on fake pool. hub.js:1148:12
15:52:04.545 [reconnect] Getting reticulum meta in 10 seconds. hub.js:1126:16
15:52:04.547 undefined
15:52:14.328 [reconnect] Getting reticulum meta. hub.js:1133:16
15:52:14.652 [reconnect] Failed to reconnect. Did not get meta for vfake version on fake pool. hub.js:1152:14
15:52:14.652 [reconnect] Reticulum deploy detected v1.0.20210323184939 on earth. hub.js:1148:12
15:52:14.652 [reconnect] Getting reticulum meta in 9 seconds. hub.js:1126:16
15:52:23.470 [reconnect] Getting reticulum meta. hub.js:1133:16
15:52:23.638 [reconnect] Reconnect in progress. Updated reticulum meta. hub.js:1156:12
15:52:23.638 Phoenix Socket URL: wss://pensive-dwarf.reticulum.io phoenix-utils.js:125:10
15:52:34.206 [reconnect] Reconnection successful. hub.js:1167:12
```
